src/core/drive/page_snapshot.ts

```diff
@@ -0,0 +1,63 @@
+import { parseHTMLDocument } from "../../util"
+import { Snapshot } from "../snapshot"
+import { expandURL } from "../url"
+import { HeadSnapshot } from "./head_snapshot"
+
+export class PageSnapshot extends Snapshot<HTMLBodyElement> {
+  static fromHTMLString(html = "") {
+    return this.fromDocument(parseHTMLDocument(html))
+  }
+
+  static fromElement(element: Element) {
+    return this.fromDocument(element.ownerDocument)
+  }
+
+  static fromDocument({ head, body }: Document) {
+    return new this(body as HTMLBodyElement, new HeadSnapshot(head))
```

Comment:
I'm surprised that this `HTMLBodyElement` casting is necessary given the parameterized type in the class definition, and that `document.body` doesn't return an `HTMLBodyElement`. Is TypeScript guarding against `document.body` returning an instance of [HTMLFrameSetElement](https://developer.mozilla.org/en-US/docs/Web/API/HTMLFrameSetElement)?

Reply:
Yeah, I wonder if it's for framesets? Or maybe some kind of XML doc that you might somehow have as a Document? Either way, `document.body` is an `HTMLElement` as far as the built-in typings are concerned.